spec/index.html

```diff
@@ -435,11 +465,47 @@ <h2>Canonicalization Algorithm</h2>
       algorithm should make the loop effectual.
     </p>
 
-    <section class="informative">
+    <section id="canon-algo-overview" class="informative">
       <h3>Overview</h3>
+
+      <p><a>URDNA2015</a> canonically labels an <a>RDF dataset</a>.
+        In URDNA2015, an RDF dataset <var>D</var>
+        is represented as a set of <a>quads</a> of the form `&lt; s, p, o, g >`
+        where the graph component `g` is empty if and only if the
+        <a>triple</a> `&lt; s, p, o >` is in the <a>default graph</a>.
+        This algorithm considers an RDF dataset to be a set of quads.
+        Two RDF datasets are considered to be the same modulo blank nodes,
+        or isomorphic, if and only if they return the same canonically labeled list of quads
+        via <a>URDNA2015</a>.</p>
+
+      <p><a>URDNA2015</a> consists of several sub-algorithms.
+        These sub-algorithms are introduced in the following sub-sections.
+        Below we give a high level summary of URDNA2015.</p>
+
+      <ol>
+        <li><strong>Initialization</strong>.
+          Initialize the state needed for the rest of the algorithm
+          using <a href="#canon-state" class="sectionRef"></a>.</li>
+        <li><strong>Compute first degree hashes</strong>.
+          Compute the first degree hash for each blank node in the dataset using <a href="#hash-1d-quads" class="sectionRef"></a>.</li>
+        <li><strong>Canonically label unique nodes</strong>.
+          Assign canonical identifiers via <a href="#issue-identifier" class="sectionRef"></a>,
+          in lexicographical order, to each blank node whose first degree hash is unique.</li>
```

Comment:
How is `lexicographical order` defined?  If I understand this doc correctly as it stands, I think including this definition is important, especially because we're working with global data and users, so need to consider the complications inherent in i18n and L10n (i.e., internationalization and localization, or internationalisation and localisation, depending on your locale ... noting also that [W3C uses US English](https://www.w3.org/Guide/manual-of-style/#Spelling)).

((tangent ... it's too bad that [`Append ",spell" to a W3C URI to invoke W3C's spell checker`](https://www.w3.org/Guide/manual-of-style/#Spelling) can't be brought to bear on GitHub-hosted documents.  I wonder whether W3C's tech magicians can provide the dictionary on which that checker must be based for use on GitHub and/or local docs and/or etc?))

Reply:
The [Unicode Collation Algorithm 15.0](https://unicode.org/reports/tr10/) defines a sorting algorithm, which can be pretty complex for encodings other than UTF-8.

Although I've always understood the meaning of "Lexicographical order" as clear (there is a [Wikipedia entry on it](https://en.wikipedia.org/wiki/Lexicographic_order), we can certainly define the term and reference Unicode ordering.

Note that we used the term "lexicographical order" and similar in JSON-LD API with no apparent misunderstanding.

Reply:
> [@gkellogg] 
> The [Unicode Collation Algorithm 15.0](https://unicode.org/reports/tr10/) defines a sorting algorithm, which can be pretty complex for encodings other than UTF-8.

Given that we're already going to require a number of transformations, I think requiring other encodings to be converted to UTF-8 is not unreasonable.  Having skimmed the Unicode sorting algorithm, I'm afraid it will be pretty complex even for UTF-8 encoded data.  (I certainly don't fully grasp it yet!)  Their discussion of [Deterministic Sorting](https://unicode.org/reports/tr10/#Deterministic_Sorting) seems particularly important for implementers of our work.

> [@gkellogg] 
> there is a [Wikipedia entry on [Lexicographical order]](https://en.wikipedia.org/wiki/Lexicographic_order)

Indeed there is ... and its second paragraph starts with —
> [**Wikipedia**] 
> [There are several variants and generalizations of the lexicographical ordering.](https://en.wikipedia.org/wiki/Lexicographic_order#:~:text=There%20are%20several%20variants%20and%20generalizations%20of%20the%20lexicographical%20ordering.)

— which is reinforced by the [several See Also items](https://en.wikipedia.org/wiki/Lexicographic_order#See_also) —

> [**Wikipedia**] 
> * [Collation](https://en.wikipedia.org/wiki/Collation)
> * [Kleene–Brouwer order](https://en.wikipedia.org/wiki/Kleene%E2%80%93Brouwer_order)
> * [Lexicographic preferences](https://en.wikipedia.org/wiki/Lexicographic_preferences)
> * [Lexicographic order topology on the unit square](https://en.wikipedia.org/wiki/Lexicographic_order_topology_on_the_unit_square)
> * [Lexicographic ordering in tensor abstract index notation](https://en.wikipedia.org/wiki/Abstract_index_notation#Braiding)
> * [Lexicographically minimal string rotation](https://en.wikipedia.org/wiki/Lexicographically_minimal_string_rotation)
> * [Long line (topology)](https://en.wikipedia.org/wiki/Long_line_(topology))
> * [Lyndon word](https://en.wikipedia.org/wiki/Lyndon_word)
> * [Star product](https://en.wikipedia.org/wiki/Star_product), a different way of combining partial orders
> * [Orders on the Cartesian product of totally ordered sets](https://en.wikipedia.org/wiki/Total_order#Orders_on_the_Cartesian_product_of_totally_ordered_sets)

> [@gkellogg] 
> Note that we used the term "lexicographical order" and similar in JSON-LD API with no apparent misunderstanding.

This the sort of thing that tends not to get raised as an issue until two people with different interpretations of the meaning of a term trade data and discover their unexpected differences with varying severity of impact.  I regret that I was not involved sufficiently with the JSON-LD API work to have caught this as an issue.

Reply:
> > [@gkellogg]
> > The [Unicode Collation Algorithm 15.0](https://unicode.org/reports/tr10/) defines a sorting algorithm, which can be pretty complex for encodings other than UTF-8.
> 
> Given that we're already going to require a number of transformations, I think requiring other encodings to be converted to UTF-8 is not unreasonable. Having skimmed the Unicode sorting algorithm, I'm afraid it will be pretty complex even for UTF-8 encoded data. (I certainly don't fully grasp it yet!) Their discussion of [Deterministic Sorting](https://unicode.org/reports/tr10/#Deterministic_Sorting) seems particularly important for implementers of our work.

I expect `String.sort()` to do the right thing on most platforms. It's beyond our scope to exhaustively explore problem areas that belong to the Unicode specs.

See if the definitions I added adequately satisfy your concerns. I really think that we could go overboard trying to specify what lexicographical ordering of Unicode strings means.

Reply:
Also, note that sorting is done after conversion to N-Quads, which requires quads to be encoded in UTF-8, which simplifies the collation problem.

Reply:
> I expect `String.sort()` as implemented on most platforms to do the right thing on most platforms.

I haven't looked at what the `Unicode Collation Algorithm 15.0` does yet -- but we should definitely make sure that whatever we're specifying matches existing behavior (or is very close and easy to implement w/o performance issues) in languages.

I think, ideally, we'd just be sorting based on Unicode code points. Unfortunately, I think JavaScript compares on code units instead -- and since it uses UTF-16, there may be a need to adjust for cases with surrogate pairs (where the first code unit is `>= 0xD800` or something like that). There's some discussion of sorting adjustments for UTF-16 here:

https://icu-project.org/docs/papers/utf16_code_point_order.html

I'm guessing Java also sorts by default using UTF-16 code units since, IIRC, it represents strings internally that way as well. I'm guessing other languages would sort by UTF-8 code units, but my understanding is that this sort order is the same as sorting by code points.

I'd expect that it would be a mistake to specify that the sort is by UTF-16 code units, so languages that do so natively might need a patch for a certain range. Unless the majority of languages do sorting that way already. Perhaps those languages could internally convert each N-Quad string to UTF-8 byte arrays and then sort by byte for byte -- or take the approaches in that link or something similar. I'm not sure what would be the most efficient solution is, but hopefully we can avoid a significant performance hit.

Reply:
Well, it looks like you can specify `ducet` (Default Unicode Collation Element Table (DUCET)) as the locale to use the Unicode Collation Algorithm in JavaScript ... if it's supported:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Locale/collation

Unfortunately, it also looks like it's perhaps not supported in browsers (well, not in Firefox anyway, if MDN docs are accurate).

Reply:
Also, for our particular use case, it's important that we don't sort based on locale, but rather, ensure that the sort order is the same regardless of locale. This points to using code points or code units to me.

IMO, our requirements are:

1. Consistent sort order no matter the locale.
2. Speed.
3. Simplicity to implement.
4. Matching (or getting a close as possible) to existing implementations.

Reply:
Unicode collation is not helpful. It is locale sensitive and you must know the locale. It is not total.

> https://unicode.org/reports/tr10/#Common_Misperceptions
> 
> "Swedish and German share most of the same characters, for example, but have very different sorting orders."
> "Collation is not a property of strings."
> "Collation order is not preserved when comparing sort keys generated from different collation sequences."

For @dlongley points 1-4, the conclusion is one of codepoint sorting (logically convert to 21 bits values, sort) or 
a code unit for a specific code unit choice.

SPARQL is (strictly) codepoint (the abstract character) ordering.

[Section 17.3](https://www.w3.org/TR/sparql11-query/#OperatorMapping) says use collation `http://www.w3.org/2005/xpath-functions/collation/codepoint` (credit to Eric Prud'hommeaux).

Reply:
I also suspect we want to compare using code points. The only reason we'd choose code units would be if most languages already natively sort by UTF-16 code units, but I think this is unlikely.

For those languages that don't offer some fast, native support for code point sorting, but instead compare natively using UTF-16 code units, I imagine the input could be scanned just once to look for surrogate pairs. Only if any were detected would special comparison code need to be used.

This discussion is somewhat split between here and #18 -- I recommend we say in this PR that lexicographical order refers to Unicode code point order and then continue the discussion over there. My guess is that existing implementations already sort by Unicode code point order or are *very close* and just need a slight adjustment for some data because they compare using UTF-16 code units today.

Reply:
Does DUCET canonicalize codepoints? (https://unicode.org/reports/tr10/#Contractions_DUCET)

So different ways of writing the same character (e.g. combining characters, [Unicode Normalization](https://unicode.org/reports/tr15/)) have the same weight but then two different xsd:strings are not totally ordered if one contains the character as one codepoint and the other used two (the second being a combining character).

Not simple any more!